### PR TITLE
Use a custom loss func to encourage complete routes

### PR DESCRIPTION
**Pull request by rparrett**
This seems to be doing stuff, but that stuff doesn't include generating good models.

Maybe I should add this also as a `custom_metric` and do some hyperparam tuning.

### CODE REVIEW

**Comment by rparrett** [comment]
This is now producing good models after https://github.com/rparrett/kilter_brain_gen/commit/d75dfb69109104344d4ce1f6e2caf521be020b77, but it's not clear at all whether this is worth doing.

Anecdotally, I still saw one route in the 30 I looked at with a missing end hold.

I would be curious to run the same sort of code as a `custom_metric` or as a filter on the training data.

**Comment by mparrett** [comment]
This is interesting! Could you share the train/eval loss curves?

A couple of thoughts off the top of my head:

1. Custom metric, yes... logging the penalties separately will make it easier to see how it evolves and correlates with other training metrics.
2. Make sure the penalties are scaled appropriately relative to the core loss. As a starting point, I'd make sure the penalties aren't contributing more than 50% to the loss. But I'm guessing we want this to be smaller. We might want to introduce a hyperparameter to scale the strength of its contribution. One approach would be to start with something like 20% relative to the normal loss and work our way up. I think the standard cross entropy loss is 0..1.
3. Possibly introduce the penalty only after certain number of epochs (this would be another hyperparam to play with); later when we understand this better we could also try building up the various penalties gradually.

**Comment by mparrett** [comment]
FYI : Was able to check out pr#2, update my venv, and start training.

**Comment by mparrett** [comment]
Trying out some of these ideas here: https://github.com/mparrett/kilter_brain_gen/commit/4c77893333412b5564022dc924c20f4ea8a9e9cf

**Comment by rparrett** [comment]
Okay, very cool. Had an immediate face-palm reaction to seeing the word "padding." Will attempt to understand how the token counting works soon, that looks neat.

**Comment by mparrett** [comment]
+1 run the same code on the training data. We really shouldn't be training on bad examples, unless we treat them specially as negative samples, which would require some work. Best to filter them out, I bet.

One weird thing we could try that I've seen suggested is inverting or augmenting the loss in the custom_loss handler for examples that exceed a certain penalty threshold.

**Comment by mparrett** [comment]
Yeah, that occurred to me later than expected, too. I'm going to share an
earlier revision of the batch token counting that's a little easier to
understand. I would hate to have to read this as is 6 months from now. I'm
not even sure the max/clamp version is correct, given the result of the
test run from last night.

**Comment by mparrett** [comment]
Added your penalty code to the dataset preprocessing and filtered anything with penalty > 0:

```
train
Filter: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 57519/57519 [00:00<00:00, 72117.36 examples/s]
Removed 91 of 57519 examples
test
Map: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 19173/19173 [00:00<00:00, 37456.85 examples/s]
Filter: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 19173/19173 [00:00<00:00, 77411.05 examples/s]
Removed 33 of 19173 examples
```

Training results pending...

**Comment by rparrett** [comment]
>  I would hate to have to read this as is 6 months from now. I'm not even sure the max/clamp version is correct

Eh, that part seemed reasonable at a glance, but I would like to do more reading on one hot encoding / various tensor methods and actually understand how stuff like this is working:
```python
one_hot = F.one_hot(predictions, num_classes=len(self.start_hold_mask)).float()
one_hot = one_hot * non_padding_mask.unsqueeze(-1)
start_holds = torch.sum(one_hot * self.start_hold_mask, dim=(1, 2))
```

**Comment by mparrett** [comment]
Changes in the below commit are working decently well for me, and I'm guessing it would be possible to swap back in the `torch.sum` / `clamp` versions if it is worth it. I think `torch.where` isn't horrible, though:

https://github.com/mparrett/kilter_brain_gen/commit/7b3c3e0d27b3c00a1f0d9819dad0168b6ab56a2f

> I would like to do more reading on one hot encoding / various tensor method

Yeah, good to understand the tensor magic!

**Comment by mparrett** [comment]
Kicked off a training run with https://github.com/mparrett/kilter_brain_gen/commit/f40b05f06f84752bc2a9685b4a91ca632fe6b279 and both train loss and eval loss are dropping so far. Prior to setting `self.penalty_alpha = 1e-1` this was not the case due to quick overfitting of the training set.
